NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System.Collections.Generic;
+using System.Reflection;
+using System.Reflection.Emit;
+using HarmonyLib;
+using NitroxClient.GameLogic;
+using NitroxClient.MonoBehaviours;
+using NitroxModel.DataStructures;
+using NitroxModel.Helper;
+
+namespace NitroxPatcher.Patches.Dynamic
+{
+    /// <summary>
+    /// Prevents sonar from turning off automatically for the player that isn't currently piloting the Cyclops.
+    /// </summary>
+    public class CyclopsSonarButton_Update_Patch : NitroxPatch, IDynamicPatch
+    {
+        public static readonly MethodInfo TARGET_METHOD = Reflect.Method((CyclopsSonarButton t) => t.Update());
+
+        private static readonly OpCode INJECTION_OPCODE = OpCodes.Ldsfld;
+        private static readonly object INJECTION_OPERAND = Reflect.Field(() => Player.main);
+
+        private static NitroxId currentCyclopsId;
+
+        public static void Prefix(CyclopsSonarButton __instance)
+        {
+            currentCyclopsId = NitroxEntity.GetId(__instance.subRoot.gameObject);
+        }
+
+        public static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions)
+        {
+            Validate.NotNull(INJECTION_OPERAND);
+            
+            /* Normally in the Update()
+             * if (Player.main.GetMode() == Player.Mode.Normal && this.sonarActive)
+		     * {
+		     *     this.TurnOffSonar();
+		     * }
+             * this part will be changed into:
+             * if (CyclopsSonarButton_Update_Patch.ShouldTurnOff() && Player.main.GetMode() == Player.Mode.Normal && this.sonarActive)
+             */
+            List<CodeInstruction> codeInstructions = new List<CodeInstruction>(instructions);
+            CodeInstruction callInstruction = new(OpCodes.Call, Reflect.Method(() => ShouldTurnoff()));
+            CodeInstruction brInstruction = new(OpCodes.Brfalse);
+            for (int i = 0; i < codeInstructions.Count; i++)
+            {
+                CodeInstruction instruction = codeInstructions[i];
+
+                if (instruction.opcode.Equals(INJECTION_OPCODE) && instruction.operand.Equals(INJECTION_OPERAND))
+                {
+                    // The second line after the current instruction should be a Brtrue, we need its operand to have the same jump label for our brfalse
+                    CodeInstruction nextBr = codeInstructions[i + 2];
+                    Validate.IsTrue(nextBr.opcode.Equals(OpCodes.Brtrue), "Looks like subnautica code has changed. Update jump offset!");
```

Comment:
This (constant validation statements in general (like at the start of the transpiler)) should be checked in the test and not at runtime.